#169258630 User Registration

william000000 opened the pull request:
#### What does this PR do?
Users should be able to sign up and receive descriptive signup/Registration validation errors or success.
#### Description of Task to be completed?
- Add validations
- Add tests
- Create user tables
- Use JSON Web token
- Swagger documentation 
#### How should this be manually tested?
- You have to clone the repo `git clone https://github.com/andela/team-odd-bn-backend.git`
- Then After the repo cloned `cd project-folder` and run `npm i` to install all dependencies
- Go to my branch `git checkout ft-signup-and-receive-descriptive-and-validation-169258630`
- Run app by `npm run dev-start`

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#169258621](https://www.pivotaltracker.com/story/show/169258630)
#### Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/40672143/68593464-30e6d700-049e-11ea-9f1f-46680a859dff.png)
![image](https://user-images.githubusercontent.com/40672143/68663071-b4f49980-0546-11ea-99e7-4a2f497b0774.png)




#### Questions:
N/A

julietezekwe committed:
Merge pull request #12 from andela/ch-setup-docker-169258626

william000000 committed:
Merge https://github.com/andela/team-odd-bn-backend into develop

hezronkimutai reviewed (commented):
Nice work willy, Just squash the commits.

victkarangwa commented on src/routes/api/users.js:
```diff
@@ -14,8 +17,8 @@ const router = require('express').Router();
  *      responses:
- *        "200":
+ *        "201":
  *          description: A user schema
```
@william000000 would you mind documenting signup properly?

william000000 replied:
@victorkarangwa4  Let me correct it

ivymwende commented on src/controllers/users.controller.js:
```diff
@@ -0,0 +1,43 @@
+      if (error) {
+        throw new Error(error.details[0].message);
+      }
+      const hashedPassword = await bcrypt.hash(req.body.password, 10);
```
@Willy I think this should be in a helper function since it is reusable code

william000000 replied:
The feedback implemented

ivymwende commented on src/controllers/users.controller.js:
```diff
@@ -0,0 +1,43 @@
+      const datas = await users.create({
+        firstName, lastName, email, password: hashedPassword
+      });
+      const token = jwt.sign({
```
@Willy also I think JWT authentication should be a helper function since it is reusable. Also, this is violation of the separation of concerns principle

william000000 replied:
The feedback implemented

ivymwende commented on src/controllers/users.controller.js:
```diff
@@ -0,0 +1,43 @@
+
+      if (isUserExist) return res.status(409).json({ status: 409, error: 'User already exist' });
+
+      const datas = await users.create({
```
I'd probably name this newUser as opposed to datas?

ivymwende commented on src/index.js:
```diff
@@ -1,15 +1,18 @@
 
 const app = express();
-const port = process.env.PORT || 8000;
+const port = process.env.PORT || 5000;
```
@william000000 Why port 5000?

ivymwende commented on src/routes/api/index.js:
```diff
@@ -1,18 +1,7 @@
+import users from './users';
 
-router.use("/", require("./users"));
+const router = require('express').Router();
```
@william000000 I think you can destructure router from express to as to follow ES6 principles like this `import { Router } from 'express';`

ivymwende commented on src/tests/mockdata.js:
```diff
@@ -0,0 +1,12 @@
+import dotenv from 'dotenv';
```
@william000000 to ensure our work is neater we can add this in a mockdata folder since we may have multiple mockdata files for other features.

ivymwende commented on src/helpers/index.js:
```diff
@@ -0,0 +1,15 @@
+/* eslint-disable require-jsdoc */
```
@william000000 I think we have a validation folder? This can go inside the validation folder rather than in the helpers folder

william000000 replied:
Feedback Implemented

julietezekwe commented on .eslintrc.json:
```diff
@@ -13,6 +13,7 @@
       "consistent-return": 0,
       "no-param-reassign": 0,
       "comma-dangle": 0,
+      "no-console": 0,
```
Please revert this. it helps check to make sure you do not push console.log

william000000 committed:
ft(signup): create user
- Add validations
- Add tests
- Add swagger documentation
- Create user database
[Start #169258630]

julietezekwe commented on .hound.yml:
```diff
@@ -1,3 +1,4 @@
 eslint:
   enabled: true
   config_file: .eslintrc
+  ignore_file: .eslintignore
```
End of line

hezronkimutai replied:
end of line

julietezekwe commented on src/database/migrations/20191108085351-create-users.js:
```diff
@@ -30,4 +33,4 @@ module.exports = {
     return queryInterface.dropTable('users');
   }
-};
+};
```
End of line

julietezekwe commented on src/database/models/users.js:
```diff
@@ -0,0 +1,69 @@
+ *           email: fname.lname@andela.com
+ *           password: yourpassword
+ *
+ */
```
End of line

hezronkimutai commented on package.json:
```diff
@@ -4,16 +4,23 @@
-    "test": "NODE_ENV=testing nyc mocha ./src/tests/*.js  --timeout 200000 --exit",
-    "coverage": "nyc npm test  && npm run coveralls-coverage"
+    "dev": "export NODE_ENV=testing && mocha src/tests/*.js --exit ",
+    "test": "NODE_ENV=testing npm run dropTables && NODE_ENV=testing npm run db-migrate && NODE_ENV=testing nyc --reporter=text --reporter=html --require @babel/polyfill --require @babel/register mocha ./src/tests/*.js  --timeout 200000 --exit",
```
I think you should remove one of this.

julietezekwe commented on src/middlewares/ValidateUser.js:
```diff
@@ -0,0 +1,38 @@
+ * @export
+ * @class Validations
+ */
+class Validations {
```
Make class name the same as the file name

julietezekwe commented on src/routes/api/index.js:
```diff
@@ -1,18 +1,7 @@
 
-router.use("/", require("./users"));
+const router = express.Router();
+router.use('/', users);
```
This will make all endpoints go to users. `/auth` for authentication and `/user/ for other user routes is better. Also, the route files should be descriptive as such `authRoute` and `userRoute.

julietezekwe commented on src/routes/api/users.js:
```diff
@@ -1,11 +1,17 @@
+import validateUser from '../../middlewares/ValidateUser';
+import isUserExist from '../../middlewares/findUser';
+
 const router = require('express').Router();
```
Since we are having different routers in different files, this file shoud be `authRouter` while this variable should be called `authRouter` as well, instead of just router.

julietezekwe commented on src/routes/index.js:
```diff
@@ -1,9 +1,11 @@
 const swaggerOption = require('../utils/swaggerOptions');
-router.use('/api', require('./api'));
 
+router.use('/api/', api);
```
Also, some versioning will be ok, we can have `api/v1/` instead of just `/api/`

julietezekwe commented on src/tests/user.test.js:
```diff
@@ -0,0 +1,82 @@
+import chai from 'chai';
```
Please rename this file to userTest

julietezekwe commented on src/controllers/users.controller.js:
```diff
@@ -0,0 +1,50 @@
+import dotenv from 'dotenv';
```
Please rename this file to UserController

william000000 replied:
Yeah I did it

julietezekwe reviewed (changes requested):
1. Please follow the right naming convention, Class names should be Same as the file name and should be PascalCase. Function names should be camelCase.
2. The method of naming where you user.controller or user.test is wrong. Please use UserController instead
3. The route names too are not descriptive.

Functionalities
1. The validation error only says `"\"firstName\" is required"` even when firstname is available. It does not display other required fields

Kindly make the changes and revert.
